dcore/state.py

```diff
@@ -120,6 +121,10 @@ def dump(self):
             self.diagnostic_data = defaultdict(partial(defaultdict, list))
 
         if (next(self.dumpcount) % self.output.dumpfreq) == 0:
+
+            if 'Courant' in self.diagnostic_fields.diagnostic_field_dict.keys():
+                self.diagnostic_fields.Courant(self)
+
```

Comment:
Note, if you find yourself writing this pattern:

```python
    def field(self, mesh):
        if hasattr(self, "_field"):
            return self._field
        self._field = Function(FunctionSpace(mesh, "DG", 0), name=self.name)
        return self._field
```

Over and over again, we can add a memoize decorator to simplify it to:

```python
@memoize
def field(self, mesh):
    return Function(FunctionSpace(mesh, "DG", 0))
```